ethcontract-mock/src/details.rs

```diff
@@ -0,0 +1,456 @@
+        let address: Address = args.arg();
+        let block: Option<BlockNumber> = args.block_number_opt();
+        args.done();
+
+        let block = block.unwrap_or(BlockNumber::Latest);
+        let state = self.state.lock().unwrap();
+        let transaction_count = match block {
+            BlockNumber::Latest | BlockNumber::Pending => {
+                state.nonce.get(&address).copied().unwrap_or(0)
+            }
+            BlockNumber::Earliest => 0,
+            _ => panic!(
+                "mock node does not support returning transaction count for specific block number"
+            ),
+        };
+        Self::ok(&U256::from(transaction_count))
+    }
+
+    fn gas_price(&self, args: Parser) -> Result<Value, Error> {
+        args.done();
+
+        let state = self.state.lock().unwrap();
+        Self::ok(&U256::from(state.gas_price))
+    }
+
+    fn estimate_gas(&self, mut args: Parser) -> Result<Value, Error> {
+        let _request: CallRequest = args.arg();
+        let _block: Option<BlockNumber> = args.block_number_opt();
+        args.done();
+
+        // TODO:
+        //
+        // We could look up contract's method, match an expectation,
+        // and see if the expectation defines gas price.
+        //
+        // So, for example, this code:
+        //
+        // ```
+        // contract
+        //     .expect_method(signature)
+        //     .with(matcher)
+        //     .gas(100);
+        // ```
+        //
+        // Indicates that call to the method with the given signature
+        // requires 100 gas.
+        //
+        // When estimating gas, we'll check all expectation as if we're
+        // executing a method, but we won't mark any expectation as fulfilled.
+
+        Self::ok(&U256::from(1))
+    }
+
+    fn send_transaction(&self, mut args: Parser) -> Result<Value, Error> {
+        let _request: TransactionRequest = args.arg();
+        args.done();
+
+        // TODO:
+        //
+        // We could support signing if user adds accounts with their private
+        // keys during mock setup.
+
+        panic!("mock node can't sign transactions, use offline signing with private key");
+    }
+
+    fn send_raw_transaction(&self, mut args: Parser) -> Result<Value, Error> {
+        let raw_tx: Bytes = args.arg();
+        args.done();
+
+        let mut state = self.state.lock().unwrap();
+
+        let tx = verify(&raw_tx.0, state.chain_id);
+
+        let nonce = state.nonce.entry(tx.from).or_insert(0);
+        if *nonce != tx.nonce.as_u64() {
+            panic!(
+                "nonce mismatch for account {:#x}: expected {}, actual {}",
+                tx.from,
+                tx.nonce.as_u64(),
+                nonce
+            );
+        }
+        *nonce += 1;
+
+        let contract = state.contract(tx.to);
+
+        let hash = tx.hash;
+
+        let status = contract.process_tx(tx);
+
+        state.block += 1;
+
+        let receipt = TransactionReceipt {
```

Comment:
I'm not sure if it's a good idea to leave most values empty here. Maybe we should check how a "normal" tx receipt looks like and use default hash values instead of None values?